llm_studio/src/utils/modeling_utils.py

```diff
@@ -90,15 +97,29 @@ def save_checkpoint(model: torch.nn.Module, path: str, cfg: Any):
         Dictionary with all the keys to save
     """
 
-    model = unwrap_model(model)
-
-    if hasattr(cfg.training, "lora") and cfg.training.lora:
-        model.backbone.save_pretrained(path)
-
-    checkpoint = {"model": model.state_dict()}
-
-    if path is not None:
-        torch.save(checkpoint, os.path.join(path, "checkpoint.pth"))
+    if cfg.environment.use_deepspeed:
+        if path is not None:
+            # gather model params from all ranks
+            if hasattr(cfg.training, "lora") and cfg.training.lora:
```

Comment:
should this only be for rank 0?

Reply:
For deepspeed, saving checkpoint requires to be called in all ranks for gathering every parameters of the model.

Reply:
also lora?

Reply:
> How exactly does the sharding work with LORA? Are all layers sharded? Would it be useful/possible to keep the LORA layers on a single gpu?

The method of sharding is no different in lora training and non-lora training. Moreover, I believe there is no need to make all lora layers on a single GPU, otherwise it would make that GPU the bottleneck of the entire training.

Therefore, when saving the model, regardless of whether it's lora training or not, it's necessary to call save_checkpoint on all ranks.